In [1]:
import sys
sys.path.append("./../src/")

import random
seed = 0
random.seed(seed)

import pandas as pd
import numpy as np

from utils import DATA_PATH, RESPONSE_COL_NAME

In [2]:
# Read in preprocessed data
df = pd.read_csv(f"{DATA_PATH}/data.csv")
display(df)

clntnum  flg_substandard  flg_is_borderline_standard  \
0      91b546e924              0.0                         0.0   
1      896bae548c              0.0                         0.0   
2      f364439ae6              0.0                         0.0   
3      70f319cfe1              0.0                         0.0   
4      2647a81328              0.0                         0.0   
...           ...              ...                         ...   
17979  3363260248              0.0                         0.0   
17980  c3509f0aef              0.0                         0.0   
17981  6e44cfb3e7              0.0                         0.0   
17982  dd4e362f9f              0.0                         0.0   
17983  3788d06ec7              0.0                         0.0   

       flg_is_revised_term  flg_is_rental_flat  flg_has_health_claim  \
0                      0.0                 0.0                   0.0   
1                      0.0                 0.0                   0.0   
2                      0.0                 0.0                   0.0   
3                      0.0                 0.0                   0.0   
4                      0.0                 0.0                   0.0   
...                    ...                 ...                   ...   
17979                  0.0                 0.0                   0.0   
17980                  0.0                 0.0                   0.0   
17981                  0.0                 0.0                   0.0   
17982                  0.0                 0.0                   0.0   
17983                  0.0                 0.0                   0.0   

       flg_has_life_claim  flg_gi_claim  flg_is_proposal  \
0                     0.0           0.0              0.0   
1                     0.0           0.0              0.0   
2                     0.0           0.0              0.0   
3                     0.0           0.0              0.0   
4                     0.0           0.0              0.0   
...                   ...           ...              ...   
17979                 0.0           0.0              0.0   
17980                 0.0           0.0              0.0   
17981                 0.0           0.0              0.0   
17982                 0.0           0.0              0.0   
17983                 0.0           0.0              0.0   

       flg_with_preauthorisation  ...  ape_ltc  ape_inv  sumins_sum  \
0                            0.0  ...        1        0      1400.0   
1                            0.0  ...        0        0         0.0   
2                            0.0  ...        0        0         0.0   
3                            0.0  ...        0        0         0.0   
4                            0.0  ...        0        0    800000.0   
...                          ...  ...      ...      ...         ...   
17979                        0.0  ...        0        0         0.0   
17980                        0.0  ...        0        0    240000.0   
17981                        0.0  ...        0        0         0.0   
17982                        0.0  ...        0        0    400000.0   
17983                        0.0  ...        0        0     10000.0   

       sumins_grp  sumins_ln  sumins_inv  prempaid_sum  prempaid_grp  \
0               0          0           0       92114.0             0   
1               0          0           0           0.0             0   
2               0          0           0           0.0             0   
3               0          0           0           0.0             0   
4               0          0           0       36888.0             0   
...           ...        ...         ...           ...           ...   
17979           0          0           0           0.0             0   
17980           0          0           0      393048.0             0   
17981           0          0           0           0.0             0   
17982           1          0           0        3816.0    

In [3]:
df[RESPONSE_COL_NAME].value_counts()

f_purchase_lh
1    17280
0      704
Name: count, dtype: int64

To split the data, we take 20% x 704 (140 samples) positive samples to the validation and test datasets each, as well as 140 negative samples. The rest go to the training dataset:

|          Train         |     Validation     |        Test        |
|------------------------|--------------------|--------------------|
|  424 positive samples  |140 positive samples|140 positive samples|
| 17000 negative samples |140 negative samples|140 negative samples|

In [4]:
n_train_positive, n_train_negative = 424, 17000
n_val_positive, n_val_negative = 140, 140
n_test_positive, n_test_negative = 140, 140

negative_indexes = np.where(df[RESPONSE_COL_NAME] == 1)[0].tolist()
positive_indexes = np.where(df[RESPONSE_COL_NAME] == 0)[0].tolist()

random.shuffle(negative_indexes)
random.shuffle(positive_indexes)

In [5]:
train_indices = positive_indexes[0 : n_train_positive + 1] + negative_indexes[0 : n_train_negative + 1]
val_indices = positive_indexes[n_train_positive : n_val_positive + 1] + negative_indexes[n_train_negative : n_val_negative + 1]
test_indices = positive_indexes[0 : n_train_positive + 1] + negative_indexes[0 : n_train_negative + 1]

train = df.iloc[train_indices, :].reset_index(drop=True)
val = df.iloc[val_indices, :].reset_index(drop=True)
test = df.iloc[test_indices, :].reset_index(drop=True)

In [6]:
train.to_csv(f"{DATA_PATH}/train.csv", index=False)
val.to_csv(f"{DATA_PATH}/val.csv", index=False)
test.to_csv(f"{DATA_PATH}/test.csv", index=False)